File: backend/app/models/order.py
from __future__ import annotations
import enum
from sqlalchemy import Column, DateTime, Float, ForeignKey, String, Boolean
from sqlalchemy.orm import relationship
from app.db.base_class import Base
from app.models.address import Address
from app.utils.utils import date_format, tz_now
from .user import User
from .payment import Payment
from .product import Product
from random import randint

def order_number_gen():
    numbers = "".join(str(i) for i in range(10))
    return "".join(numbers[randint(0, 9)] for i in range(5))

class OrderStatus(str, enum.Enum):
    DELIVERED = "DELIVERED"
    CANCELLED = "CANCELLED"
    PENDING = "PENDING"
    ACTIVE = "ACTIVE"


class OrderStatusEvent(str, enum.Enum):
    CREATED = "CREATED"
    CANCELLED = "CANCELLED"
    CONFIRMED = "CONFIRMED"
    DELIVERED = "DELIVERED"
    OTHER = "OTHER"
    PROCESSED = "PROCESSED"


class Order(Base):
    order_number = Column(String(5), default=order_number_gen())
    created_at = Column(DateTime(), default=tz_now(), nullable=False)
    user_id = Column(ForeignKey(f"{User.__tablename__}.id"), nullable=False)
    payment_id = Column(ForeignKey(f"{Payment.__tablename__}.id"))
    user = relationship(User.__tablename__)
    payment = relationship(Payment.__tablename__)
    order_items = relationship("OrderItem")
    shipment_cost = Column(Float(), default=0, nullable=False)
    detail = relationship("OrderDetail", foreign_keys="OrderDetail.order_id")
    status: OrderStatus = Column(
        String(10), nullable=False, default=OrderStatus.PENDING
    )
    delivery_address_id = Column(
        ForeignKey(f"{Address.__tablename__}.id"), nullable=False
    )
    delivery_address = relationship(Address.__tablename__)

    @property
    def total(self):
        return sum([item.quantity * item.product.price for item in self.order_items])

    @property
    def created_at_format(self):
        return date_format(self.created_at)

    @property
    def quantity(self):
        return len(self.order_items)


class OrderItem(Base):
    product_id = Column(ForeignKey(f"{Product.__tablename__}.id"), nullable=False)
    product = relationship(
        f"{Product.__tablename__}",
    )
    order_id = Column(
        ForeignKey(f"{Order.__tablename__}.id", ondelete="CASCADE"),
        nullable=False,
    )
    quantity = Column(Float(precision=2), nullable=False)
    delivered: bool = Column(Boolean(), nullable=False, default=False)


class OrderDetail(Base):
    order_id = Column(ForeignKey(f"{Order.__tablename__}.id", ondelete="CASCADE"))
    message = Column(String(1000), nullable=False)
    created_at = Column(DateTime(), default=tz_now(), nullable=False)
    order_item_id = Column(ForeignKey(f"{Order.__tablename__}.id"))
    event = Column(String(30), nullable=False, default="OTHER")

File: market/lib/data/models/order/order.dart
// ignore_for_file: constant_identifier_names

import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:market/data/models/address/address.dart';
import 'package:market/data/models/payment/payment.dart';
import 'package:market/data/models/product/product.dart';
import 'package:market/data/models/user/user.dart';
part 'order.g.dart';
part 'order.freezed.dart';

enum OrderStatus { PENDING, ACTIVE, CANCELLED, DELIVERED, INITIAL }

enum OrderDetailEvent {
  CREATED,
  CANCELLED,
  CONFIRMED,
  DELIVERED,
  OTHER,
  PROCESSED
}

extension Str on OrderStatus {
  String toStr() {
    switch (this) {
      case OrderStatus.ACTIVE:
        return "Active";
      case OrderStatus.PENDING:
        return "Pending";
      case OrderStatus.CANCELLED:
        return "Cancelled";
      case OrderStatus.DELIVERED:
        return "Delivered";
      default:
        return "Initial";
    }
  }
}

@Freezed(fromJson: true, copyWith: true)
class Order with _$Order {
  factory Order({
    @JsonKey(name: "created_at") required DateTime createdAt,
    int? id,
    @JsonKey(defaultValue: OrderStatus.PENDING) required OrderStatus status,
    User? user,
    @JsonKey(name: "delivery_address") required Address address,
    @JsonKey(name: "order_items") required List<OrderItem> items,
    Payment? payment,
    @JsonKey(name: "shipping_cost", defaultValue: 0)
        required double shippingCost,
    required List<OrderDetail> detail,
  }) = _Order;
  Order._();
  double get cost =>
      items.fold(0, (prev, item) => prev + item.quantity * item.product.price);

  factory Order.fromJson(json) => _$$_OrderFromJson(json);
}

@Freezed(fromJson: true, copyWith: true)
class OrderItem with _$OrderItem {
  factory OrderItem({
    required Product product,
    required double quantity,
    @JsonKey(defaultValue: false) required bool delivered,
    int? id,
  }) = _OrderItem;
  factory OrderItem.fromJson(json) => _$$_OrderItemFromJson(json);
}

@Freezed()
class OrderItemCreate with _$OrderItemCreate {
  OrderItemCreate._();
  factory OrderItemCreate({
    required int productId,
    required double quantity,
  }) = _OrderItemCreate;
  Map toJson() => {
        "product_id": productId,
        "quantity": quantity,
      };
}

@Freezed(fromJson: true, copyWith: true)
class OrderDetail with _$OrderDetail {
  factory OrderDetail({
    @JsonKey(name: "order_item_id") int? orderItemId,
    required String message,
    @JsonKey(name: "created_at") required DateTime createdAt,
    required OrderDetailEvent event,
  }) = _OrderDetail;
  factory OrderDetail.fromJson(json) => _$$_OrderDetailFromJson(json);
}

extension Ex on OrderDetail {
  String get icon {
    switch (event) {
      case OrderDetailEvent.CONFIRMED:
        return "assets/confirmed.png";
      case OrderDetailEvent.CREATED:
        return "assets/placed.png";
      case OrderDetailEvent.CANCELLED:
        return "assets/cancel.png";
      case OrderDetailEvent.PROCESSED:
        return "assets/processes.png";
      case OrderDetailEvent.DELIVERED:
        return "assets/ready_to_pickup.png";
      default:
        return "assets/hint.png";
    }
  }
}
